src/number_theory/legendre_symbol/quadratic_reciprocity.lean

```diff
@@ -141,213 +133,240 @@ lemma legendre_sym_eq_neg_one_iff_not_one {a : ℤ} (ha : (a : zmod p) ≠ 0) :
-  cases (legendre_sym_eq_one_or_neg_one p q hqp0) with h h,
-  { simp only [h, eq_self_iff_true, true_iff, mul_one] at this ⊢,
-    exact this, },
-  { simp only [h, mul_neg, mul_one] at this ⊢,
-    rw eq_neg_of_eq_neg this.symm, },
+  have h := quadratic_char_neg_two (ne_of_eq_of_ne (ring_char_zmod_n p) hp),
+  rw [card p] at h,
+  rw [legendre_sym],
+  exact_mod_cast h,
 end
 
-lemma exists_sq_eq_prime_iff_of_mod_four_eq_three (hp3 : p % 4 = 3)
-  (hq3 : q % 4 = 3) (hpq : p ≠ q) : is_square (q : zmod p) ↔ ¬ is_square (p : zmod q) :=
-have h1 : ((p / 2) * (q / 2)) % 2 = 1,
-  from nat.odd_mul_odd
-    (by rw [← mod_mul_right_div_self, show 2 * 2 = 4, from rfl, hp3]; refl)
-    (by rw [← mod_mul_right_div_self, show 2 * 2 = 4, from rfl, hq3]; refl),
+/-- `-2` is a square modulo an odd prime `p` iff `p` is congruent to `1` or `3` mod `8`. -/
+lemma exists_sq_eq_neg_two_iff : is_square (-2 : zmod p) ↔ p % 8 = 1 ∨ p % 8 = 3 :=
 begin
-  have hp_odd : p ≠ 2 := by { by_contra, simp [h] at hp3, norm_num at hp3, },
-  have hq_odd : q ≠ 2 := by { by_contra, simp [h] at hq3, norm_num at hq3, },
-  have hpq0 : ((p : ℤ) : zmod q) ≠ 0 := by exact_mod_cast prime_ne_zero q p (ne.symm hpq),
-  have hqp0 : ((q : ℤ) : zmod p) ≠ 0 := by exact_mod_cast prime_ne_zero p q hpq,
-  have := quadratic_reciprocity p q hp_odd hq_odd hpq,
-  rw [neg_one_pow_eq_pow_mod_two, h1, pow_one] at this,
-  rw [(by norm_cast : (p : zmod q) = (p : ℤ)), (by norm_cast : (q : zmod p) = (q : ℤ)),
-       ← legendre_sym_eq_one_iff _ hpq0, ← legendre_sym_eq_one_iff _ hqp0],
-  cases (legendre_sym_eq_one_or_neg_one q p hpq0) with h h,
-  { simp only [h, eq_self_iff_true, not_true, iff_false, one_mul] at this ⊢,
-    simp only [this],
-    norm_num, },
-  { simp only [h, neg_mul, one_mul, neg_inj] at this ⊢,
-    simp only [this, eq_self_iff_true, true_iff],
-    norm_num, },
+  rw [finite_field.is_square_neg_two_iff, card p],
+  have h₁ := prime.mod_two_eq_one_iff_ne_two.mpr hp,
+  rw [← mod_mod_of_dvd p (by norm_num : 2 ∣ 8)] at h₁,
+  have h₂ := mod_lt p (by norm_num : 0 < 8),
+  revert h₂ h₁,
+  generalize hm : p % 8 = m, unfreezingI {clear_dependent p},
+  dec_trivial!,
 end
 
+end values
+
+section reciprocity
+
+/-!
+### The Law of Quadratic Reciprocity
+-/
+
+variables {p q : ℕ} [hpp : fact p.prime] [hqp : fact q.prime]
+include hpp hqp
+
+/-- The Law of Quadratic Reciprocity: if `p` and `q` are distinct odd primes, then
+`(q / p) * (p / q) = (-1)^((p-1)(q-1)/4)`. -/
+theorem quadratic_reciprocity (hp : p ≠ 2) (hq : q ≠ 2) (hpq : p ≠ q) :
+  legendre_sym q p * legendre_sym p q = (-1) ^ ((p / 2) * (q / 2)) :=
+begin
+  have hp₁ := (prime.eq_two_or_odd hpp.1).resolve_left hp,
+  have hq₁ := (prime.eq_two_or_odd hqp.1).resolve_left hq,
+  have hq₂ := ne_of_eq_of_ne (ring_char_zmod_n q) hq,
+  have hpq₁ : (p : zmod q) ≠ 0 :=
+  (mt (nat_coe_zmod_eq_zero_iff_dvd p q).mp
+        $ mt (prime_dvd_prime_iff_eq hqp.1 hpp.1).mp hpq.symm),
```

Comment:
```suggestion
  have hpq₁ : (p : zmod q) ≠ 0 := prime_ne_zero q p hpq.symm,
```
could be inlined as well.

Reply:
OK. This also means that now, `hpp` and `hqp` are used only once, so I dropped the names and use `fact.out`.